plugins/Amplifier/Amplifier.cpp

```diff
@@ -65,17 +65,17 @@ bool AmplifierEffect::processAudioBuffer(sampleFrame* buf, const fpp_t frames)
 	const float d = dryLevel();
 	const float w = wetLevel();
 
-	const ValueBuffer* volumeBuf = m_ampControls.m_volumeModel.valueBuffer();
-	const ValueBuffer* panBuf = m_ampControls.m_panModel.valueBuffer();
-	const ValueBuffer* leftBuf = m_ampControls.m_leftModel.valueBuffer();
-	const ValueBuffer* rightBuf = m_ampControls.m_rightModel.valueBuffer();
+	const auto* volumeBuf = m_ampControls.m_volumeModel.valueBuffer();
+	const auto* panBuf = m_ampControls.m_panModel.valueBuffer();
+	const auto* leftBuf = m_ampControls.m_leftModel.valueBuffer();
+	const auto* rightBuf = m_ampControls.m_rightModel.valueBuffer();
 
 	for (fpp_t f = 0; f < frames; ++f)
 	{
-		const float volume = (volumeBuf ? volumeBuf->value(f) : m_ampControls.m_volumeModel.value()) * 0.01f;
-		const float pan = (panBuf ? panBuf->value(f) : m_ampControls.m_panModel.value()) * 0.01f;
-		const float left = (leftBuf ? leftBuf->value(f) : m_ampControls.m_leftModel.value()) * 0.01f;
-		const float right = (rightBuf ? rightBuf->value(f) : m_ampControls.m_rightModel.value()) * 0.01f;
+		const float volume = (volumeBuf ? volumeBuf->at(f % volumeBuf->size()) : m_ampControls.m_volumeModel.value()) * 0.01f;
```

Comment:
I'm curious if the modulo was covering for some bug in ValueBuffer before.

Reply:
which bug?

Reply:
The issue is that this is hiding out of bounds array access. The question is why is this needed.

Reply:
I just implemented the Old behaviour from ValueBuffer which was:

```c++

float ValueBuffer::value(int offset) const
{
	return at(offset % length());
}
```

Reply:
Right, and the question is why was modulo being used here.

Reply:
I have no idea

Reply:
do you think we should remove the modulo ?

Reply:
I think we should definitely look into it.

Reply:
@DomClark what do you think?

Reply:
Wrapping around seems like a bizarre solution to handling out-of-bounds access. Clamping the index would make more sense, but I think we should find out why (if at all) it is accessed out-of-bounds, and add an assertion if it shouldn't be.

Reply:
should be fine now

Reply:
@Snowiiii can you report on what happened when you removed the modulo/clamp?

Reply:
> @Snowiiii can you report on what happened when you removed the modulo/clamp?

everything works just fine
![image](https://github.com/LMMS/lmms/assets/71594357/1cd654fb-cb0d-4ee7-a6a5-829b77667b60)

The only issue i currently no lines in the piano editor will be rendered
![image](https://github.com/LMMS/lmms/assets/71594357/557f5beb-eaa4-4931-9756-97aa35529f6f)
but im not sure if its my fault or some commit broke something, using the latest https://github.com/LMMS/lmms/commit/9dd7f4dde456efa95586a0c4744255039f965c78 the problem does not exist

Reply:
Yeah the notes not rendering was fixed. Can you load and play an intensive project like Krem Kaakkuja with no modulo/clamp? If so, then I think we can remove the modulo/clamp.

Reply:
> Yeah the notes not rendering was fixed. Can you load and play an intensive project like Krem Kaakkuja with no modulo/clamp? If so, then I think we can remove the modulo/clamp.

works without any problems.
![image](https://github.com/LMMS/lmms/assets/71594357/5580e4d0-f660-48a1-80a5-a7b7e19f518c)

Reply:
@DomClark I think we can remove the clamp, too, now. What do you think?

Reply:
@PhysSong for additional opinion :)

Reply:
I think you should use static code analysis to check every occurrence where the value function is used. In all these cases it must be assured that no out of bounds read is possible. If in doubt, I would suggest to add an assert before such calls and then make tests where this assert is being tested.

Reply:
I think we should leave it an runtime check